### PR TITLE
feat(tt-core): migrate OpenCode ingestion from JSON files to SQLite

### PR DESCRIPTION
**Pull request by sjawhar**
## Summary

OpenCode recently changed its storage format from JSON files (`~/.local/share/opencode/storage/`) to a SQLite database (`~/.local/share/opencode/opencode.db`). This PR updates the session ingestion logic accordingly.

### Changes

- **`tt-core/src/opencode.rs`** — Complete rewrite: reads from SQLite via `json_extract()` queries instead of walking JSON files. Uses `prepare_cached` for per-session queries, read-only connection with busy timeout for safe concurrent access.
- **`tt-core/Cargo.toml`** — Added `rusqlite.workspace = true`
- **`tt-core/src/session.rs`** — Added `SessionError::Database` variant
- **`tt-cli/src/commands/ingest.rs`** — Updated path from `storage/` dir to `opencode.db`
- **`tt-core/AGENTS.md`** — Updated documentation

### Key design decisions

- **Read-only + `SQLITE_OPEN_NO_MUTEX`** — Safe for concurrent reads while OpenCode is running
- **`json_valid()` guards** in SQL queries — Gracefully handles malformed JSON data columns
- **Graceful degradation** — Missing/corrupt db returns empty vec with warning (matches prior behavior for missing dirs)
- **Per-session tool count** — Single `SELECT COUNT(*)` per session instead of per-message part queries

### Testing

- 16 tests ported from JSON fixtures to SQLite fixtures (same assertions)
- Added `test_scan_corrupt_db` and `test_malformed_message_data` edge cases
- All 299 tests pass across all crates, clippy clean

## Post-Deploy Monitoring & Validation

No additional operational monitoring required: this is a local CLI tool, not a deployed service.

### CODE REVIEW

File: crates/tt-core/src/opencode.rs [hunk 1]
```diff
@@ -1,367 +1,359 @@
...
-    if !session_dir.exists() {
-        return Ok(Vec::new());
-    }
+fn count_tool_calls(conn: &Connection, session_id: &str) -> Result<i32, SessionError> {
+    let mut tool_stmt = conn.prepare_cached(
+        "SELECT COUNT(*) FROM part \
+         WHERE session_id = ?1 AND json_valid(data) \
+         AND json_extract(data, '$.type') = 'tool'",
+    )?;
+    let tool_count: i64 = tool_stmt.query_row([session_id], |row| row.get::<_, i64>(0))?;
```
Comment (sjawhar) [inline crates/tt-core/src/opencode.rs]: This counts **all** tool parts for the session regardless of parent message role. The old JSON implementation only counted tool parts while iterating **assistant** messages:

```rust
MessageRole::Assistant => {
    let count = parts.iter().filter(|p| p.part_type == PartType::Tool).count();
    tool_call_count = tool_call_count.saturating_add(...);
}
```

If OpenCode ever attaches tool parts to non-assistant messages, this will diverge.

Fix: either join `part → message` and add `AND json_extract(message.data, '$.role') = 'assistant'`, or compute tool counts inline while iterating assistant messages in `collect_message_stats`.

File: crates/tt-core/src/opencode.rs [hunk 2]
```diff
@@ -1,367 +1,359 @@
...
+                    if stats.user_message_timestamps.len() < MAX_USER_MESSAGE_TIMESTAMPS {
+                        if let Some(ts) = unix_ms_to_datetime(created_ms) {
+                            stats.user_message_timestamps.push(ts);
+                        }
+                    }
+                }
+            }
+            "assistant" => {
+                stats.assistant_message_count = stats.assistant_message_count.saturating_add(1);
+                stats.last_message_time = Some(created_ms);
```
Comment (sjawhar) [inline crates/tt-core/src/opencode.rs]: Old code set `last_message_time` unconditionally for *every* message before the role match:

```rust
for msg in &messages {
    last_message_time = Some(msg.time.created); // all roles
    // ... then match on role
}
```

Here, messages with roles other than `"user"`/`"assistant"` silently skip without updating `last_message_time`, so they won't extend `end_time`.

Fix: move `stats.last_message_time = Some(created_ms);` before the `match` (after the `let Some(role)` guard) to restore parity.

File: crates/tt-core/src/opencode.rs [hunk 3]
```diff
@@ -1,367 +1,359 @@
...
 }
 
-/// Read and deserialize all JSON files from a directory.
-fn read_json_files<T: serde::de::DeserializeOwned>(dir: &Path) -> Vec<T> {
-    let Ok(entries) = fs::read_dir(dir) else {
-        return Vec::new();
+pub fn scan_opencode_sessions(db_path: &Path) -> Result<Vec<AgentSession>, SessionError> {
+    let flags = OpenFlags::SQLITE_OPEN_READ_ONLY | OpenFlags::SQLITE_OPEN_NO_MUTEX;
+    let conn = match Connection::open_with_flags(db_path, flags) {
+        Ok(conn) => conn,
```
Comment (sjawhar) [inline crates/tt-core/src/opencode.rs]: Nit: worth a brief comment here explaining why `NO_MUTEX` is safe — the connection is only used from a single thread. If someone later re-adds rayon parallelism (the old code used `par_iter`), this flag would be unsafe without the context.

**Review by sjawhar — COMMENTED** [review]
## Code Review

Overall this is a clean, well-tested migration. Graceful degradation, `prepare_cached` usage, `json_valid()` guards, and the test suite are all solid. Two behavioral divergences from the old implementation worth closing before merge.

### Important

1. **Tool call counting may diverge from old behavior** — see inline comment on `count_tool_calls()`.

2. **`end_time` ignores non-user/assistant message roles** — see inline comment on the `_ => {}` arm in `collect_message_stats`.

### Minor

3. **`SQLITE_OPEN_NO_MUTEX`** needs a comment explaining the single-thread assumption — see inline.
4. **Docstring in `ingest.rs:348`** still says "session directories" but OpenCode is now a db file. Suggest: *"Scans Claude Code session directories and the OpenCode SQLite database..."*

File: crates/tt-core/src/opencode.rs (hunk 1 above)
Comment (sjawhar) [reply crates/tt-core/src/opencode.rs]: Fixed. Joined `part → message` with `json_extract(m.data, '$.role') = 'assistant'` filter so only assistant-message tool parts are counted.

File: crates/tt-core/AGENTS.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+## Session Scanning (`session.rs` + `opencode.rs`)
+
+Parses Claude Code (`~/.claude/projects/`) JSONL session files and OpenCode (`~/.local/share/opencode/opencode.db`) SQLite database.
+
+### Key Types
+
+- `AgentSession` — parsed session: `session_id`, `source`, `parent_session_id`, `session_type`, `project_path`, `start_time`, `end_time`, `message_count`, `user_prompts`, etc.
+- `SessionSource` — enum: `Claude` | `OpenCode`
+- `SessionType` — enum: `User` | `Agent` | `Subagent` (inferred from session_id format)
```
[the remark this answers is not in the document]
Comment (sjawhar) [reply crates/tt-core/AGENTS.md]: Fixed in AGENTS.md — clarified that `session.rs` uses session_id heuristics while `opencode.rs` uses `parent_id` presence.
